File: src/apps/Andriod/app/src/main/java/com/dkellycollins/triggerfinger/data/daos/impl/BulletDao.java
package com.dkellycollins.triggerfinger.data.daos.impl;

import android.os.Bundle;

import com.dkellycollins.triggerfinger.data.daos.IBulletDao;
import com.dkellycollins.triggerfinger.data.entity.IBullet;
import com.dkellycollins.triggerfinger.data.entity.impl.Bullet;
import com.dkellycollins.triggerfinger.data.model.IVector;

import java.io.Serializable;
import java.util.HashMap;

/**
 * Standard implementation of IBulletDao.
 */
public class BulletDao extends BaseEntityDao<Bullet> implements IBulletDao {

    public BulletDao() {
        super("BulletDao");
    }

    @Override
    public Iterable<IBullet> retrieve() {
        return (Iterable<IBullet>) (Iterable<?>) _store.values();
    }

    @Override
    public IBullet retrieve(int id) {
        return _store.get(id);
    }

    @Override
    public int create(int collidableId, IVector velocity) {
        int id = getNextId();
        Bullet bullet = new Bullet(id, collidableId, velocity);

        _store.put(id, bullet);

        return id;
    }

    @Override
    public void delete(int id) {
        _store.remove(id);
    }
}

File: src/apps/Android-Kotlin/app/src/main/kotlin/com/dkellycollins/triggerfinger/data/daos/impl/BulletDao.kt
package com.dkellycollins.triggerfinger.data.daos.impl

import android.os.Bundle

import com.dkellycollins.triggerfinger.data.daos.IBulletDao
import com.dkellycollins.triggerfinger.data.entity.IBullet
import com.dkellycollins.triggerfinger.data.entity.impl.Bullet
import com.dkellycollins.triggerfinger.data.model.IVector

import java.io.Serializable
import java.util.HashMap

/**
 * Standard implementation of IBulletDao.
 */
class BulletDao : BaseEntityDao<Bullet>("BulletDao"), IBulletDao {

    override fun retrieve(): Iterable<IBullet> {
        return _store.values
    }

    override fun retrieve(id: Int): IBullet {
        return _store[id]
    }

    override fun create(collidableId: Int, velocity: IVector): Int {
        val id = nextId
        val bullet = Bullet(id, collidableId, velocity)

        _store.put(id, bullet)

        return id
    }

    override fun delete(id: Int) {
        _store.remove(id)
    }
}